Add events back to NodeController

```
I0331 00:11:32.928517   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.929362   17044 nodecontroller.go:214] Registered node in registry: node0
I0331 00:11:32.929423   17044 nodecontroller.go:214] Registered node in registry: node1
I0331 00:11:32.929459   17044 nodecontroller.go:222] Successfully registered all nodes
I0331 00:11:32.929882   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.930213   17044 nodecontroller.go:214] Registered node in registry: node0
I0331 00:11:32.930680   17044 nodecontroller.go:214] Registered node in registry: node1
I0331 00:11:32.930734   17044 nodecontroller.go:222] Successfully registered all nodes
I0331 00:11:32.930972   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.931689   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:32.932004   17044 nodecontroller.go:217] Error registering node node0, retrying: Create error.
E0331 00:11:32.932072   17044 nodecontroller.go:217] Error registering node node1, retrying: Create error.
I0331 00:11:32.933263   17044 nodecontroller.go:214] Registered node in registry: node0
I0331 00:11:32.933317   17044 nodecontroller.go:214] Registered node in registry: node1
I0331 00:11:32.933349   17044 nodecontroller.go:222] Successfully registered all nodes
I0331 00:11:32.933757   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.934069   17044 nodecontroller.go:214] Registered node in registry: node0
I0331 00:11:32.934150   17044 nodecontroller.go:214] Registered node in registry: node1
I0331 00:11:32.934184   17044 nodecontroller.go:222] Successfully registered all nodes
I0331 00:11:32.934592   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:32.934912   17044 nodecontroller.go:217] Error registering node node0, retrying: Create error.
I0331 00:11:32.934982   17044 nodecontroller.go:214] Registered node in registry: node1
E0331 00:11:32.939045   17044 nodecontroller.go:217] Error registering node node0, retrying: Create error.
I0331 00:11:32.943914   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.944202   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.944589   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.945200   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.945487   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.945836   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.946880   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.947489   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.948529   17044 nodecontroller.go:266] Create node in registry: node1
I0331 00:11:32.948791   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.949254   17044 nodecontroller.go:280] Delete node from registry: node1
I0331 00:11:32.951303   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.951939   17044 nodecontroller.go:266] Create node in registry: node1
I0331 00:11:32.953224   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.955448   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.963457   17044 nodecontroller.go:280] Delete node from registry: node1
I0331 00:11:32.964477   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:32.971069   17044 nodecontroller.go:280] Delete node from registry: node1
I0331 00:11:32.972389   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.983307   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.983619   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.984532   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
I0331 00:11:32.984979   17044 nodecontroller.go:91] No api server defined - no events will be sent to API server.
E0331 00:11:32.985171   17044 nodecontroller.go:482] error getting instance addresses for node0: unable to query IP address.
E0331 00:11:32.985219   17044 nodecontroller.go:482] error getting instance addresses for node1: unable to query IP address.
I0331 00:11:32.988499   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:32.989255   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:32.989354   17044 nodecontroller.go:335] Can't collect information for node node1: Not Implemented
E0331 00:11:32.990060   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:32.990155   17044 nodecontroller.go:335] Can't collect information for node node1: Not Implemented
I0331 00:11:32.990747   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:32.991488   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:33.005489   17044 nodecontroller.go:503] Can't get ip address of node node0: lookup node0: no such host
I0331 00:11:33.005739   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:33.006366   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:33.006475   17044 nodecontroller.go:503] Can't get ip address of node node0: lookup node0: no such host
I0331 00:11:33.007014   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:33.007588   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:33.007674   17044 nodecontroller.go:503] Can't get ip address of node node0: lookup node0: no such host
I0331 00:11:33.008035   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:33.015280   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:33.031118   17044 nodecontroller.go:503] Can't get ip address of node node0: lookup node0: no such host
==================
WARNING: DATA RACE
Read by goroutine 66:
  reflect.Value.String()
      /usr/local/go/src/reflect/value.go:1601 +0x8e
  fmt.(*pp).printReflectValue()
      /usr/local/go/src/fmt/print.go:867 +0x4fe9
  fmt.(*pp).printValue()
      /usr/local/go/src/fmt/print.go:836 +0x5a6
  fmt.(*pp).printReflectValue()
      /usr/local/go/src/fmt/print.go:918 +0x5785
  fmt.(*pp).printValue()
      /usr/local/go/src/fmt/print.go:836 +0x5a6
  fmt.(*pp).printReflectValue()
      /usr/local/go/src/fmt/print.go:976 +0x18a5
  fmt.(*pp).printArg()
      /usr/local/go/src/fmt/print.go:798 +0x52e
  fmt.(*pp).doPrintf()
      /usr/local/go/src/fmt/print.go:1183 +0x2ec9
  fmt.Sprintf()
      /usr/local/go/src/fmt/print.go:203 +0x85
  testing.(*common).Errorf()
      /usr/local/go/src/testing/testing.go:357 +0x5d
  github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller.TestSyncProbedNodeStatusEvictPods()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/nodecontroller_test.go:1076 +0x25eb
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133
Previous write by goroutine 71:
  github.com/GoogleCloudPlatform/kubernetes/pkg/client.(*FakeEvents).Create()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/fake_events.go:35 +0x196
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.recordEvent()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:149 +0x431
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.func·001()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:120 +0x281
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.func·003()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:206 +0x14e
Goroutine 66 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:485 +0xe0
  main.main()
      github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/_test/_testmain.go:122 +0x381
Goroutine 71 (running) created at:
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:208 +0x185
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartRecordingToSink()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:136 +0x219
  github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller.NewNodeController()
      github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/_test/_obj_test/nodecontroller.go:88 +0x19c
  github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller.TestSyncProbedNodeStatusEvictPods()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/nodecontroller_test.go:1067 +0x2038
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133
==================
==================
WARNING: DATA RACE
Read by goroutine 66:
  reflect.Value.IsNil()
      /usr/local/go/src/reflect/value.go:940 +0xc0
  fmt.getField()
      /usr/local/go/src/fmt/print.go:282 +0xd2
  fmt.(*pp).printReflectValue()
      /usr/local/go/src/fmt/print.go:918 +0x5723
  fmt.(*pp).printValue()
      /usr/local/go/src/fmt/print.go:836 +0x5a6
  fmt.(*pp).printReflectValue()
      /usr/local/go/src/fmt/print.go:976 +0x18a5
  fmt.(*pp).printArg()
      /usr/local/go/src/fmt/print.go:798 +0x52e
  fmt.(*pp).doPrintf()
      /usr/local/go/src/fmt/print.go:1183 +0x2ec9
  fmt.Sprintf()
      /usr/local/go/src/fmt/print.go:203 +0x85
  testing.(*common).Errorf()
      /usr/local/go/src/testing/testing.go:357 +0x5d
  github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller.TestSyncProbedNodeStatusEvictPods()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/nodecontroller_test.go:1076 +0x25eb
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133
Previous write by goroutine 71:
  github.com/GoogleCloudPlatform/kubernetes/pkg/client.(*FakeEvents).Create()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/fake_events.go:35 +0x196
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.recordEvent()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:149 +0x431
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.func·001()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:120 +0x281
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.func·003()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:206 +0x14e
Goroutine 66 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:485 +0xe0
  main.main()
      github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/_test/_testmain.go:122 +0x381
Goroutine 71 (running) created at:
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:208 +0x185
  github.com/GoogleCloudPlatform/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartRecordingToSink()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/record/event.go:136 +0x219
  github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller.NewNodeController()
      github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/_test/_obj_test/nodecontroller.go:88 +0x19c
  github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller.TestSyncProbedNodeStatusEvictPods()
      /home/travis/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller/nodecontroller_test.go:1067 +0x2038
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133
==================
I0331 00:11:33.032235   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:33.032821   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:33.032907   17044 nodecontroller.go:503] Can't get ip address of node node0: lookup node0: no such host
I0331 00:11:33.033099   17044 nodecontroller.go:87] Sending events to api server.
E0331 00:11:33.036468   17044 nodecontroller.go:335] Can't collect information for node node0: Not Implemented
E0331 00:11:33.036571   17044 nodecontroller.go:503] Can't get ip address of node node0: lookup node0: no such host
--- FAIL: TestSyncProbedNodeStatusEvictPods (0.03s)
	nodecontroller_test.go:1076: time out waiting for deleting pods, expected [], got [{Action:get-event Value:node0.13d06cc923707dac}]
I0331 00:11:33.038672   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.039598   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.040409   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.051886   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.056438   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.057784   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.059663   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.060474   17044 nodecontroller.go:87] Sending events to api server.
I0331 00:11:33.066690   17044 nodecontroller.go:87] Sending events to api server.
FAIL
coverage: 78.5% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller	0.279s
```

/cc @ddysher

--- a/pkg/cloudprovider/controller/nodecontroller.go
+++ b/pkg/cloudprovider/controller/nodecontroller.go
@@ -26,8 +26,10 @@ import (
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
 	apierrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
 	"github.com/golang/glog"
 )
@@ -56,6 +58,7 @@ type NodeController struct {
 	nodes                   []string
 	kubeClient              client.Interface
 	kubeletClient           client.KubeletClient
+	recorder                record.EventRecorder
 	registerRetryCount      int
 	podEvictionTimeout      time.Duration
 	deletingPodsRateLimiter util.RateLimiter
@@ -104,6 +107,14 @@ func NewNodeController(
 	nodeMonitorGracePeriod time.Duration,
 	nodeStartupGracePeriod time.Duration,
 	nodeMonitorPeriod time.Duration) *NodeController {
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(api.EventSource{Component: "controllermanager"})
+	if kubeClient != nil {
+		glog.Infof("Sending events to api server.")
+		eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+	} else {
+		glog.Infof("No api server defined - no events will be sent to API server.")
+	}
 	return &NodeController{
 		cloud:                   cloud,
 		matchRE:                 matchRE,
@@ -111,6 +122,7 @@ func NewNodeController(
 		staticResources:         staticResources,
 		kubeClient:              kubeClient,
 		kubeletClient:           kubeletClient,
+		recorder:                recorder,
 		registerRetryCount:      registerRetryCount,
 		podEvictionTimeout:      podEvictionTimeout,
 		deletingPodsRateLimiter: deletingPodsRateLimiter,
@@ -298,6 +310,19 @@ func (nc *NodeController) PopulateAddresses(nodes *api.NodeList) (*api.NodeList,
 	return nodes, nil
 }
 
+func (nc *NodeController) recordNodeEvent(node *api.Node, event string) {
+	ref := &api.ObjectReference{
+		Kind:      "Node",
+		Name:      node.Name,
+		UID:       types.UID(node.Name),
+		Namespace: "",
+	}
+	glog.V(2).Infof("Recording %s event message for node %s", event, node.Name)
+	// TODO: This requires a transaction, either both node status is updated
+	// and event is recorded or neither should happen, see issue #6055.
+	nc.recorder.Eventf(ref, event, "Node %s is now %s", node.Name, event)
+}
+
 // For a given node checks its conditions and tries to update it. Returns grace period to which given node
 // is entitled, state of current and last observed Ready Condition, and an error if it ocured.
 func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, api.NodeCondition, *api.NodeCondition, error) {
@@ -480,6 +505,17 @@ func (nc *NodeController) MonitorNodeStatus() error {
 				if nc.deletingPodsRateLimiter.CanAccept() {
 					nc.deletePods(node.Name)
 				}
+			}
+
+			// Report node events.
+			if readyCondition.Status == api.ConditionTrue && lastReadyCondition.Status != api.ConditionTrue {
+				nc.recordNodeEvent(node, "ready")
+			}
+			if readyCondition.Status == api.ConditionFalse && lastReadyCondition.Status != api.ConditionFalse {
+				nc.recordNodeEvent(node, "not_ready")
+			}
+			if readyCondition.Status == api.ConditionUnknown && lastReadyCondition.Status != api.ConditionUnknown {
+				nc.recordNodeEvent(node, "unknown")
 			}
 		}
 	}

